src/txrequest.cpp

```diff
@@ -0,0 +1,724 @@
+            // announcement.
+            if (info.m_candidate_ready > 0) {
+                assert(info.m_candidate_best + info.m_requested == 1);
+            }
+
+            // If there is both a CANDIDATE_READY and a CANDIDATE_BEST announcement, the CANDIDATE_BEST one must be
+            // at least as good (equal or lower priority) as the best CANDIDATE_READY.
+            if (info.m_candidate_ready && info.m_candidate_best) {
+                assert(info.m_priority_candidate_best >= info.m_priority_best_candidate_ready);
+            }
+
+            // No txhash can have been announced by the same peer twice.
+            std::sort(info.m_peers.begin(), info.m_peers.end());
+            assert(std::adjacent_find(info.m_peers.begin(), info.m_peers.end()) == info.m_peers.end());
+        }
+    }
+
+    void PostGetRequestableSanityCheck(std::chrono::microseconds now) const
+    {
+        for (const Announcement& ann : m_index) {
+            if (ann.IsWaiting()) {
+                // REQUESTED and CANDIDATE_DELAYED must have a time in the future (they should have been converted
+                // to COMPLETED/CANDIDATE_READY respectively).
+                assert(ann.m_time > now);
+            } else if (ann.IsSelectable()) {
+                // CANDIDATE_READY and CANDIDATE_BEST cannot have a time in the future (they should have remained
+                // CANDIDATE_DELAYED, or should have been converted back to it if time went backwards).
+                assert(ann.m_time <= now);
+            }
+        }
+    }
+
+private:
+    //! Wrapper around Index::...::erase that keeps m_peerinfo up to date.
+    template<typename Tag>
+    Iter<Tag> Erase(Iter<Tag> it)
+    {
+        auto peerit = m_peerinfo.find(it->m_peer);
+        peerit->second.m_completed -= it->m_state == State::COMPLETED;
+        peerit->second.m_requested -= it->m_state == State::REQUESTED;
+        if (--peerit->second.m_total == 0) m_peerinfo.erase(peerit);
+        return m_index.get<Tag>().erase(it);
+    }
+
+    //! Wrapper around Index::...::modify that keeps m_peerinfo up to date.
+    template<typename Tag, typename Modifier>
+    void Modify(Iter<Tag> it, Modifier modifier)
+    {
+        auto peerit = m_peerinfo.find(it->m_peer);
+        peerit->second.m_completed -= it->m_state == State::COMPLETED;
+        peerit->second.m_requested -= it->m_state == State::REQUESTED;
+        m_index.get<Tag>().modify(it, std::move(modifier));
+        peerit->second.m_completed += it->m_state == State::COMPLETED;
+        peerit->second.m_requested += it->m_state == State::REQUESTED;
+    }
+
+    //! Convert a CANDIDATE_DELAYED announcement into a CANDIDATE_READY. If this makes it the new best
+    //! CANDIDATE_READY (and no REQUESTED exists) and better than the CANDIDATE_BEST (if any), it becomes the new
+    //! CANDIDATE_BEST.
+    void PromoteCandidateReady(Iter<ByTxHash> it)
```

Comment:
I think this function is right but I was confused for a while on the behavior if we can have concurrently two announcements, one  _BEST the other _REQUESTED. We can't but it would be better to document it as a friendly reminder.

Reply:
It says "and no REQUESTED exists", is that enough?

Reply:
Right there is an and marking the conjunction.